### PR TITLE
Safer comparisons on DCR (ESLint)

### PR DESCRIPTION
**Pull request by OllysCoding**
_tl;dr: This PR adds an ESLint rule which bans 2 things:_
1. _Ambiguous type comparisons_
2. _truthy/falsy comparisons on `number | undefined` types_

## Safer comparisons on DCR

In the past, DCR allowed any forms of type comparisons to be written. This is great for fast development, as trust is left in the developer to write & test their solutions.

As DCR has grown & scaled though, with more and more developers contributing to our platform, we need to scale our safety to ensure that code remains maintainable.

By introducing some rules around type comparisons, we can ensure that any future developers working on our platform are able to discern the intended functionality of a comparison, and avoid ambiguity or bugs being introduced where elements of truthy/fasley are forgotten.

### What is an ambiguous type comparison?

Imagine we have the following piece of code

```ts
const isInArr = (str: string | undefined) => {
    const arr = ["value1", "value2"];
    return str && arr.includes(str);
}
```

You might think from looking at that code that the return type would be a `boolean`, but actually it's `boolean | "" | undefined`.

Why? Well if we step through our comparison - `str && arr.includes(str)` - we can start to see why:

On the left hand side, we have a string which could be undefined, and are evaluating it to truthy or falsy. This gives us the cases:
**Falsy:** If the value is `undefined`, or an empty string (`""`)
**Truthy**: if the string is defined, and isn't empty, e.g `"value1"`.

Because we're using the `&&` operator, if the left hand side evaluates to falsey, it returns the left hand side. So here we're returning either `""` or `undefined`, or continuing to evaluate the right hand side of the statement. 

Luckily, the right hand side will always evaluate to `true` or `false`, but unfortunately this leaves us still with this `boolean | "" | undefined` return type.

We can see then, this find of type is not ideal, as if we were to start using this function is our code, there would be lots of really confusing types which could be hard to reason about if you weren't familiar with how the original function works.

### Why ban truthy/falsy on `number | undefined`?

An ongoing issue on our platforms are star ratings. When the guardian provides an /5 rating of a piece of media content, we rate them from **0** to 5. However this star rating doesn't exist on _all_ guardian articles, which means we get the type `number | undefined`.

We've therefor had problems in the past where we check to see if a star rating exists on a piece of content using truthy/falsy (e.g `!!starRating`). In these cases, if a piece was rated 0 stars, it wouldn't receive a star rating.

Some of these bugs still exist in DCR, as changes in this PR will highlight. 

Given the prevalence of this issue and how easy it is to introduce in this and similar cases, it makes sense to limit this type comparison specifically, meaning developers should now use `myNumber !== undefined` or `myNumber !== undefined && myNumber !== 0` in order to be explicit about what functionality we intend.

---

This PR uses `@typescript-eslint/strict-boolean-expressions'` in order to prevent these unsafe comparisons. To enable this we had to resolve ~160 errors. 

While the goal originally was only to prevent `number | undefined` expressions, it seems that the ESLint rule also prevents these 'ambiguous' comparisons outlined above. Ultimately this is probably a good thing, as it highlighted how many awkward types we had, and it was often difficult to discern what the expected functionality is.

There are some cases where an unsafe or illogical type comparison was either necessary (e.g browser modules which may not exist), or where the types were too difficult to simplify (e.g react children). In these cases we've used ignore comments & added information on why each has been ignored.

### CODE REVIEW

File: dotcom-rendering/src/model/extract-ga.ts [hunk 1]
```diff
@@ -16,8 +16,7 @@ const filterTags = (
 			[],
 		);
 
-	// eslint-disable-next-line @typescript-eslint/prefer-optional-chain -- this code is cursed
-	return (arrOfvalues && arrOfvalues.join(',')) || '';
+	return (arrOfvalues !== false && arrOfvalues.join(',')) || '';
```
Comment (mxdvl) [inline dotcom-rendering/src/model/extract-ga.ts:19]:
```suggestion
	return (Array.isArray(arrOfvalues) && arrOfvalues.join(',')) || '';
```

File: dotcom-rendering/src/web/browser/relativeTime/updateTimeElements.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export const updateTimeElement = (element: Element): void => {
 		const absoluteTime = element.getAttribute('datetime');
 		if (!absoluteTime || !relativeFormat) return;
 
-		let newTime;
+		let newTime: false | string = false;
```
Comment (mxdvl) [inline dotcom-rendering/src/web/browser/relativeTime/updateTimeElements.ts:10]:
```suggestion
		let newTime: false | undefined;
```

File: dotcom-rendering/src/web/components/InteractiveContentsBlockComponent.importable.tsx [hunk 3]
```diff
@@ -176,7 +176,7 @@ export const InteractiveContentsBlockComponent = ({
 	// -> If we stop intersecting with the current element, and its in view > show the preview element
 	// -> If we're intersecting with an element set that to the active one
 	// -> If the end element is in view, but the last element is not > show the last element
-	useEffect((): any => {
+	useEffect((): void | (() => void) => {
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/InteractiveContentsBlockComponent.importable.tsx:179]: Do you even need to specify the return type?

```suggestion
	useEffect(function observeElement() {
```

File: dotcom-rendering/src/web/components/MediaMeta.tsx [hunk 4]
```diff
@@ -126,7 +126,7 @@ export const MediaMeta = ({
 				hasKicker={hasKicker}
 			/>
 			&nbsp;
-			{!!mediaDuration && (
+			{mediaDuration !== undefined && (
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/MediaMeta.tsx:129]: What does a 0 media duration mean?

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx [hunk 5]
```diff
@@ -47,16 +47,11 @@ const ALLOWED_TAGS = [
 	'football/liverpool',
 ];
 
+// This function looks for the first tag in pageTags, that also exists in our allowlist
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:50]: JSDoc maybe?

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx [hunk 6]
```diff
@@ -66,7 +61,7 @@ const firstPopularTag = (
 	}
 
 	const firstTagInAllowedList =
-		tags.find((tag: string) => ALLOWED_TAGS.includes(tag)) ?? false;
+		tags.find((tag: string) => ALLOWED_TAGS.includes(tag)) ?? undefined;
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:64]: I think `undefined` is the return of `Array.prototype.find` already?

**Review by mxdvl — APPROVED** [review]
I like this a lot: less ambiguity!

Started reviewing individual changes, but there's quite a few bits so feel free to ignore.

File: dotcom-rendering/src/web/browser/relativeTime/updateTimeElements.ts (hunk 2 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/browser/relativeTime/updateTimeElements.ts:10]: The `timeAgo` function returns a string or false, so we can't use undefined (at least not easily)

File: dotcom-rendering/src/amp/lib/real-time-config.ts [hunk 7]
```diff
@@ -75,7 +75,7 @@ export const realTimeConfig = ({
 		notUndefined,
 	);
 
-	const options = timeoutMillis ? { timeoutMillis } : {};
+	const options = timeoutMillis !== undefined ? { timeoutMillis } : {};
```
Comment (mxdvl) [inline dotcom-rendering/src/amp/lib/real-time-config.ts:78]: Maybe simpler to just do:

```suggestion
	const options = { timeoutMillis };
```

File: dotcom-rendering/src/web/components/Card/components/CardLayout.tsx [hunk 8]
```diff
@@ -26,7 +26,7 @@ const decideDirection = (imagePosition: ImagePositionType) => {
 };
 
 const decideWidth = (minWidthInPixels?: number) => {
-	if (minWidthInPixels) {
+	if (minWidthInPixels !== undefined && minWidthInPixels !== 0) {
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/Card/components/CardLayout.tsx:29]: What does it mean to have a negative `min-width`?

```suggestion
	if (minWidthInPixels !== undefined && minWidthInPixels >= 0) {
```

File: dotcom-rendering/src/web/components/Carousel.importable.tsx [hunk 9]
```diff
@@ -563,6 +563,7 @@ export const Carousel = ({ heading, trails, onwardsSource, format }: Props) => {
 			</LeftColumn>
 			<div css={[buttonContainerStyle, prevButtonContainerStyle(format)]}>
 				<button
+					type="button"
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/Carousel.importable.tsx:566]: 👀 

Indication we might want to switch to Source buttons?

File: dotcom-rendering/src/web/components/CommentCount.importable.tsx [hunk 10]
```diff
@@ -81,7 +81,7 @@ export const CommentCount = ({
 		joinUrl(discussionApiUrl, 'discussion', shortUrlId),
 	);
 
-	if (!commentCount && commentCount !== 0) return null;
+	if (commentCount !== undefined && commentCount !== 0) return null;
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/CommentCount.importable.tsx:84]: I do not understand this logic… should it not be checking whether there are at least one comment?

```suggestion
	if (commentCount === undefined || commentCount < 1) return null;
```

File: dotcom-rendering/src/web/components/Dropdown.tsx [hunk 11]
```diff
@@ -361,16 +361,17 @@ const DropdownLink = ({ link, index }: DropdownLinkProps) => {
...
-					css={css`
-						margin-top: 12px;
-						margin-right: 8px;
-					`}
-				>
-					<NotificationBadge diameter={22} />
-				</div>
-			)}
+			{link.notifications !== undefined &&
+				link.notifications.length > 0 && (
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/Dropdown.tsx:365]: Once you start to have multiple checks, it might be better to extract its logic to a variable:

```tsx
const hasNotifications: boolean = link.notifications !== undefined && link.notifications.length > 0;

if(hasNotifications) { return <>Notifications!</> }
```

File: dotcom-rendering/src/web/components/FilterButton.importable.tsx [hunk 12]
```diff
@@ -71,7 +71,9 @@ const valueStyles = css`
 const Label = ({ value, count }: LabelProps) => (
 	<>
 		<span css={valueStyles}>{value}</span>{' '}
-		{!!count && <span css={countStyles}>({count})</span>}
+		{count !== undefined && count !== 0 && (
+			<span css={countStyles}>({count})</span>
+		)}
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/FilterButton.importable.tsx:76]: It is quite odd that this is required at all… what does a value of `0` even mean? How can you you have a selected topic without a count?

I think that it would be best to fallback to a default value of `0` for `count` and check if it’s a strictly positive integer:

```tsx
const Label = ({ value, count = 0 }: LabelProps) => (
	<>
		<span css={valueStyles}>{value}</span>{' '}
		{count > 0 && (
			<span css={countStyles}>({count})</span>
		)}
	</>
);
```

File: dotcom-rendering/src/web/components/ImageComponent.tsx [hunk 13]
```diff
@@ -265,10 +265,13 @@ export const ImageComponent = ({
...
-	const isSupported = (imageUrl: string) => {
+	const isSupported = (imageUrl: string): boolean => {
 		const supportedImages = ['jpg', 'jpeg', 'png'];
 		const extension = imageUrl.split('.').slice(-1)[0];
-		return extension && supportedImages.includes(extension.toLowerCase());
+		return (
+			extension !== undefined &&
+			supportedImages.includes(extension.toLowerCase())
+		);
 	};
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/ImageComponent.tsx:275]: Maybe the logic could be simplified with the `endsWith` method?

```ts
const isSupported = (imageUrl: string): boolean => {
	const supportedImages = ['jpg', 'jpeg', 'png'];
	return supportedImages.some((extension) => imageUrl.endsWith(`.${extension}`));
};
```

File: dotcom-rendering/src/web/components/Island.tsx [hunk 14]
```diff
@@ -44,7 +42,7 @@ const decideChildren = (
 	placeholderHeight?: number,
 ) => {
 	if (!clientOnly) return children; // Server side rendering
-	if (placeholderHeight)
+	if (placeholderHeight !== undefined && placeholderHeight !== 0)
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/Island.tsx:45]: I don’t think a negative of 0 value make any sense, here.

```suggestion
	if (placeholderHeight !== undefined && placeholderHeight > 0)
```

File: dotcom-rendering/src/web/components/RichLink.tsx [hunk 15]
```diff
@@ -296,7 +296,7 @@ export const RichLink = ({
 									{mainContributor}
 								</div>
 							)}
-							{!!starRating && (
+							{starRating !== undefined && (
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/RichLink.tsx:299]: 👏 🌟

**Review by mxdvl — COMMENTED** [review]
Worth noting that [`@guardian/libs@13` exports `isNonNullable`](https://github.com/guardian/csnx/releases/tag/%40guardian%2Flibs%4013.0.0) which can replace `item !== undefined` and `item !== null`.

File: dotcom-rendering/src/amp/lib/get-video-id.ts [hunk 16]
```diff
@@ -15,12 +15,15 @@ export const getIdFromUrl = (
...
-	].filter((tryId): tryId is string => !!tryId);
+	const ids: string[] = [];
+	if (tryQueryParam && url.query) {
+		const maybeId = new URLSearchParams(url.query).get(tryQueryParam);
+		if (maybeId) ids.push(maybeId);
+	}
+	if (tryInPath && url.pathname) {
+		const maybeId = url.pathname.split('/').pop();
+		if (maybeId) ids.push(maybeId);
+	}
```
Comment (mxdvl) [inline dotcom-rendering/src/amp/lib/get-video-id.ts:26]: I think the immutable version is superior, with a few modifications:

```ts
import { isString } from '@guardian/libs';

const url = new URL(urlString);

const ids: [
	tryQueryParam ? new URLSearchParams(url.query ?? '').get(tryQueryParam) : undefined,
	tryInPath ? url.pathname.split('/').at(-1) : undefined,
].filter(isString);
```

File: dotcom-rendering/src/web/browser/ga/ga.ts [hunk 17]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/strict-boolean-expressions, @typescript-eslint/no-unnecessary-condition --
+	We check the existence of window.ga frequently, which is required for browser safety despite the type definitions. */
```
Comment (mxdvl) [inline dotcom-rendering/src/web/browser/ga/ga.ts:2]: Should we make it optional in our type definitions?

File: dotcom-rendering/src/web/browser/ophan/index.ts [hunk 18]
```diff
@@ -8,6 +8,7 @@ export const ophan = (): Promise<void> => {
 	record({ edition: window.guardian.config.page.edition });
 
 	// Record server-side AB test variants (i.e. control or variant)
+	// eslint-disable-next-line @typescript-eslint/strict-boolean-expressions, @typescript-eslint/no-unnecessary-condition -- Client-side safety,
```
Comment (mxdvl) [inline dotcom-rendering/src/web/browser/ophan/index.ts:11]: Should we make it optional in our type definitions?

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx (hunk 5 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:50]:
```suggestion
/** This function looks for the first tag in pageTags, that also exists in our allowlist */
```

File: dotcom-rendering/src/web/server/pageTemplate.ts [hunk 19]
```diff
@@ -61,14 +61,13 @@ export const pageTemplate = ({
...
-				.join('\n');
-		}
+		return Object.entries(dataObject)
+			.map(
+				([id, value]) =>
+					`<meta ${attributeName}="${id}" content="${value}"/>`,
+			)
+			.join('\n');
+
 		return '';
```
Comment (mxdvl) [inline dotcom-rendering/src/web/server/pageTemplate.ts:71]: I think this return is not necessary, and you can even drop the curly braces altogether!

File: dotcom-rendering/src/web/components/InteractiveContentsBlockComponent.importable.tsx (hunk 3 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/InteractiveContentsBlockComponent.importable.tsx:179]: It seems if we don't we get a `Not all code paths return a value.` from TS. Guess it doesn't like the ambiguity

File: dotcom-rendering/src/web/components/MediaMeta.tsx (hunk 4 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/MediaMeta.tsx:129]: I guess you'd expect it to _not_ be 0, but don't necessarily check for it, in the same way you wouldn't expect a timestamp to be 0. Maybe we should be more explicit and check though?

File: dotcom-rendering/src/amp/lib/real-time-config.ts (hunk 7 above)
Comment (OllysCoding) [reply dotcom-rendering/src/amp/lib/real-time-config.ts:78]: This code has since been removed :)

File: dotcom-rendering/src/web/components/FilterButton.importable.tsx (hunk 12 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/FilterButton.importable.tsx:76]: Yeah I too am confused by the logic ... I didn't want. to refactor too much of edex's code, but I'll go with your suggestion :)

File: dotcom-rendering/src/web/browser/ga/ga.ts (hunk 17 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/browser/ga/ga.ts:2]: It's a bit confusing to me... window.guardian.ga is defined in our types and is optional, I think this is probably worth looking at in a new PR

File: dotcom-rendering/src/web/components/FilterButton.importable.tsx (hunk 12 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/FilterButton.importable.tsx:76]: None of these comments are blockers… but it’s hard to know what to include in a refactor. Maybe @guardian/editorial-experience can chime in directly?

File: dotcom-rendering/src/web/browser/ophan/index.ts (hunk 18 above)
Comment (OllysCoding) [reply dotcom-rendering/src/web/browser/ophan/index.ts:11]: Looking at this again I don't think we gain any safety ... updated

File: dotcom-rendering/src/web/components/ImageComponent.tsx (hunk 13 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/ImageComponent.tsx:275]: I was worried the creation of an array could be an issue, but it’s also 45% faster! 144 ns/iter v 213 ns/iter

```ts
const isSupportedSplit = (imageUrl: string): boolean => {
  const supportedImages = ["jpg", "jpeg", "png"];
  const extension = imageUrl.split(".").slice(-1)[0];
  return (
    extension !== undefined &&
    supportedImages.includes(extension.toLowerCase())
  );
};

const isSupportedEndsWith = (imageUrl: string): boolean => {
  const supportedImages = ["jpg", "jpeg", "png"];
  return supportedImages.some((extension) =>
    imageUrl.endsWith(`.${extension}`)
  );
};

Deno.bench("split", () => {
  const ok = isSupportedSplit("./some/image.jpg");
  const bad = isSupportedSplit("./some/image.svg");
  if (!ok || bad) throw Error("something went wrong");
});

Deno.bench("ends with", () => {
  const ok = isSupportedEndsWith("./some/image.jpg");
  const bad = isSupportedEndsWith("./some/image.svg");
  if (!ok || bad) throw Error("something went wrong");
});
```

File: dotcom-rendering/src/amp/lib/get-video-id.ts (hunk 16 above)
Comment (OllysCoding) [reply dotcom-rendering/src/amp/lib/get-video-id.ts]: Took a bit more modification, personally I still find the mutable version easier to read, but I'll stick with immutability since that's how it worked before!

File: dotcom-rendering/src/web/components/MediaMeta.tsx (hunk 4 above)
Comment (JamieB-gu) [reply dotcom-rendering/src/web/components/MediaMeta.tsx:129]: I think it's a valid state for the duration of some audio or a video to be zero.

**Review by JamieB-gu — COMMENTED** [review]
This is great, some big improvements here, nice one! 🚀

**Review by JamieB-gu — APPROVED** [review]

File: dotcom-rendering/src/web/components/ImageComponent.tsx [hunk 20]
```diff
@@ -265,10 +265,11 @@ export const ImageComponent = ({
 		getMaster(element.media.allImages) ??
 		getLargest(element.media.allImages);
 
-	const isSupported = (imageUrl: string) => {
+	const isSupported = (imageUrl: string): boolean => {
 		const supportedImages = ['jpg', 'jpeg', 'png'];
-		const extension = imageUrl.split('.').slice(-1)[0];
-		return extension && supportedImages.includes(extension.toLowerCase());
+		return supportedImages.some((extension) =>
+			imageUrl.endsWith(` .${extension}`),
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/web/components/ImageComponent.tsx:271]: Good catch!

File: dotcom-rendering/src/web/components/LiveBlock.tsx [hunk 21]
```diff
@@ -41,8 +41,8 @@ export const LiveBlock = ({
 	// Decide if the block has been updated or not
 	const showLastUpdated: boolean =
 		!!block.blockLastUpdatedDisplay &&
-		!!block.blockFirstPublished &&
-		!!block.blockLastUpdated &&
+		block.blockFirstPublished !== undefined &&
+		block.blockLastUpdated !== undefined &&
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/LiveBlock.tsx:45]: I think we may need to investigate how this affects preview still, but yes I generally agree :)

File: dotcom-rendering/src/web/components/TopicFilterBank.tsx [hunk 22]
```diff
@@ -116,8 +116,8 @@ export const TopicFilterBank = ({
 				Filters <span css={headlineAccentStyles(palette)}>BETA</span>
 			</div>
 			<div css={topicStyles}>
-				{hasKeyEvents ? (
-					<FilterButton
+				{keyEvents && keyEvents.length > 0 ? (
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/TopicFilterBank.tsx:119]: I think we can keep `hasKeyEvents`, here, which is a strict boolean.

```suggestion
				{hasKeyEvents ? (
```

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx [hunk 23]
```diff
@@ -47,16 +47,11 @@ const ALLOWED_TAGS = [
...
 	isPaidContent: boolean,
-) => {
-	// This function looks for the first tag in pageTags, that also exists in our allowlist
-	if (!pageTags) {
-		// If there are no page tags we will never find a match so
-		return false;
-	}
-
+): string | undefined => {
 	// The problem here is keywordIds is sometimes a string and sometimes an array of strings. Fun times.
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:55]: This is actually untrue… it’s always a `string`, fix incoming.

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx [hunk 24]
```diff
@@ -47,16 +47,11 @@ const ALLOWED_TAGS = [
 	'football/liverpool',
 ];
 
+/** This function looks for the first tag in pageTags, that also exists in our allowlist */
```
Comment (mxdvl) [inline dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:50]: This method is actually not doing what one might expect: https://github.com/guardian/dotcom-rendering/pull/1078#issuecomment-1446489263

**Review by mxdvl — COMMENTED** [review]
Discovered some issue with the `firstPopularTag` method, working on a fix

File: dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx (hunk 24 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/components/OnwardsUpper.importable.tsx:50]: Fix in https://github.com/guardian/dotcom-rendering/pull/7300
